### ISSUE TITLE
Hook manifests are applied ordered by name rather than by Kind

### ISSUE THREAD

**Issue by roobre** [helm/helm#10768]
Hey there,

While recently working on one of our charts, our team (@alvarocabanas) discovered a weird behavior on how Helm manages the order in which hooks are applied into the cluster. We noticed this after seeing that a `Job` annotated as a `pre-install` hook was getting stuck due to the `ServiceAccount` it required (and was also annotated) not being there. After some trial and error, we figured that this condition only happened when the name of the `ServiceAccount` was such that it would appear later on an alphabetically sorted list that also contained the `Job` name.

Our understanding is that, as defined here, this should not happen: https://github.com/helm/helm/blob/8ab18f7567cedffdfa5ba4d7f6abfb58efc313f8/pkg/releaseutil/kind_sorter.go#L31

Moreover the docs [say](https://helm.sh/docs/topics/charts_hooks/):
> The library sorts hooks by weight (assigning a weight of 0 by default), by resource kind and finally by name in ascending order.

Which is definitely not the behavior we were seeing.

After some troubleshooting and messing with the debugger, we found out this to be the problematic function:

https://github.com/helm/helm/blob/8ab18f7567cedffdfa5ba4d7f6abfb58efc313f8/pkg/action/hooks.go#L41-L42

Which eventually calls:

https://github.com/helm/helm/blob/8ab18f7567cedffdfa5ba4d7f6abfb58efc313f8/pkg/action/hooks.go#L115-L120

Our understanding is that the code is not doing what the `// hooke are pre-ordered by kind, so keep order stable` comment hints it is supposed to be doing. The Less function defined above will actually _override_ the Kind order if the weight is equal, resulting in a list ordered by weight, name, and finally by kind.

Our team has prepared a very naive fix for this, which essentially gets rid of the name-based sort: https://github.com/helm/helm/compare/main...newrelic-forks:fix-hook-order

As I said, this fix is very naive and maybe not ready to be merged, but we will be happy to pursue this and contribute a proper fix with this, if you can confirm this issue and guide us in the right direction.

---

Output of `helm version`:

```
version.BuildInfo{Version:"v3.8+unreleased", GitCommit:"8ab18f7567cedffdfa5ba4d7f6abfb58efc313f8", GitTreeState:"dirty", GoVersion:"go1.17.8"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"archive", BuildDate:"2022-02-18T22:31:12Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.4", GitCommit:"e6c093d87ea4cbb530a7b2ae91e54c0842d8308a", GitTreeState:"clean", BuildDate:"2022-02-16T12:32:02Z", GoVersion:"go1.17.7", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
```
Kubeadm
```

**Comment by bacongobbler** [comment]
Hooks are sorted slightly differently than standard resources. Originally Helm hooks only supported Pods, removing the need for sorting hooks by kind. This was because a hook was meant to execute a one-off job.

https://github.com/helm/helm/commit/9a2ff7802ff84fcc260c0eff940fa0af3ea91137 introduced the notion of sorting hooks by Kind if they have the same hook weight. This was introduced in Helm 3.2.

Removing the name-based sort is a no-go, as that has been the behaviour for quite some time. Removing that could break charts that rely on name-based hook ordering, such as charts that relied on its behaviour when they were written back for Helm 2.

To clarify, here's how hooks should be sorted:

1. by weight
2. within the same weight -> by kind
3. within the same kind -> by name

In other words, if I have the following list of hooks:

```
name | weight | kind
--------------------
foo  | 1      | pod
bar  | 1      | pod
zoo  | 10     | pod
bcd  | 10     | job
cde  | 10     | job
cat  | 1      | job
```

The sort order should be:

```
name | weight | kind
--------------------
bcd  | 10     | job
cde  | 10     | job
zoo  | 10     | pod
cat  | 1      | job
bar  | 1      | pod
foo  | 1      | pod
```

Hope this helps.

**Comment by roobre** [comment]
Hi @bacongobbler,

Thanks for clarifying! The behaviour you are describing is definitely what I'd expect, but not what I'm seeing on Helm 3.8 (or `master`).

To use your example, I'm seeing the following:

```
name | weight | kind
--------------------
aaa | 0 | Pod (job)
zzz | 0 | ServiceAccount (job)
```

While I'd expect to see:

```
name | weight | kind
--------------------
zzz | 0 | ServiceAccount (job)
aaa | 0 | Pod (job)
```

In my view this is a bug, as in this case the Kind ordering should have priority, and ServiceAccount is Higher than Pod in [the list](https://github.com/helm/helm/blob/8ab18f7567cedffdfa5ba4d7f6abfb58efc313f8/pkg/releaseutil/kind_sorter.go#L31
). Moreover, this will cause installations to break as the Pod will be deployed before the ServiceAccount it needs to run.

> Removing the name-based sort is a no-go, as that has been the behaviour for quite some time.

I definitely agree with this. The patch I linked was merely conceptual to demonstrate the source of the problem. If we agree that this behavior we're seeing is in fact a bug, I will be happy to iterate on the patch so it properly respects the `Weight > Kind > Name` ordering. Does that seem good to you?

**Comment by bacongobbler** [comment]
Sounds good to me
